### PR TITLE
choose-gui: init 1.3.1

### PR DESCRIPTION
**Pull request by heywoodlh**
## Description of changes

Adds [choose](https://github.com/chipsenkbeil/choose), a fuzzy finder for MacOS

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/ch/choose-gui/package.nix
Comment (superherointj) [inline pkgs/by-name/ch/choose-gui/package.nix:15]: install hooks are missing

File: pkgs/by-name/ch/choose-gui/package.nix
Comment (superherointj) [inline pkgs/by-name/ch/choose-gui/package.nix:21]:
```suggestion
  meta = {
```
Avoid nested with.

**Comment by superherointj** [comment]
Both commits needs to be reworded.

> maintainers: add heywoodlh

Should be first commit.

Second commit should be:

> choose-gui: init 1.3.1

Notice: `init` is after `:`.

File: pkgs/by-name/ch/choose-gui/package.nix
Comment (heywoodlh) [reply pkgs/by-name/ch/choose-gui/package.nix:15]: Resolved on latest commit!

**Comment by heywoodlh** [comment]
@superherointj I believe I've resolved the issues you have raised, please let me know if there's anything else you need from me.

File: pkgs/by-name/ch/choose-gui/package.nix
Comment (superherointj) [inline pkgs/by-name/ch/choose-gui/package.nix:7]: Consider using `fetchFromGitHub`. (That's the preferable option.)

File: pkgs/by-name/ch/choose-gui/package.nix
Comment (superherointj) [inline pkgs/by-name/ch/choose-gui/package.nix:24]: `meta.changelog` is missing
https://github.com/chipsenkbeil/choose/blob/master/CHANGELOG.md

**Comment by superherointj** [comment]
> @superherointj I believe I've resolved the issues you have raised, please let me know if there's anything else you need from me.

I have provided you general advice.
I don't have a Darwin host so I can't test this package myself.
I suggest you to forward it for review by darwin maintainers.

**Comment by superherointj** [comment]
`choose-gui` commit message **body** is missing:

> Release: https://github.com/chipsenkbeil/choose/releases/tag/1.3.1

(Whenever adding a new package, add the release link to the commit message. This is used by reviewers.)

**Comment by superherointj** [comment]
heywoodlh, don't get discouraged if you are having a rough start here at nixpkgs.

What happens is, you don't know the format of things yet. So you get these endless review loops and delays.
Once you get the hang of things, you start getting instant merges (which is often happens with me).
Also, once you start helping you get back and forth reviews from contributors that are alike (same ecosystem).
So things flow faster.
I multi-task over many unrelated tasks, so the feedback you are going to get might vary.
Whenever someone thinks your code is good, they're very likely to merge it.
We don't like open PRs/issue. We strive to close them all.
It is just too much for not that many contributors.
Anyway, my point is, I expect your improve to improve over time.
Just keep going. And don't mind the delays.

I recommend having a personal nixpkgs fork (for cherry-picking commits) and a nix overlay (for doing experiments and being a limbo for nixpkgs).

File: pkgs/by-name/ch/choose-gui/package.nix
Comment (heywoodlh) [reply pkgs/by-name/ch/choose-gui/package.nix:7]: I think I'm going to leave it as I have it now because it looks like the build process relies on some XCode tooling. I will investigate further tomorrow, though -- would probably be better to build the binary via Nix instead of relying on pre-built binaries.

Can `fetchFromGitHub` be used to to grab assets from a release?

File: pkgs/by-name/ch/choose-gui/package.nix
Comment (heywoodlh) [reply pkgs/by-name/ch/choose-gui/package.nix:24]: Didn't know about this field! Thanks for the pointer, fixed on the latest commit.

**Comment by heywoodlh** [comment]
@superherointj I really appreciate the encouragement, it's very helpful, especially for a noob like myself! :smile:

**Comment by heywoodlh** [comment]
> choose-gui commit message body is missing

I'm not sure I'm following. Can you provide me an example of how this looks in another package for me to replicate?

>  I suggest you to forward it for review by darwin maintainers.

How do you suggest I do this?

**Comment by superherointj** [comment]
> > choose-gui commit message body is missing
> 
> I'm not sure I'm following. Can you provide me an example of how this looks in another package for me to replicate?

Git commits have a title and a body. You just have to add to the body of commit message the line:

> Release: https://github.com/chipsenkbeil/choose/releases/tag/1.3.1

Try:

> git rebase -i HEAD~2

Alter from `pick` to `reword` the commit you want to edit the message, save+exit editor, this will re-open editor for editing the commit message.
The title of commit message is already right. You just have to paste the missing line to the next line, which is the 'body' of commit message.

**Comment by superherointj** [comment]
> > I suggest you to forward it for review by darwin maintainers.
> 
> How do you suggest I do this?

There isn't a perfect recipe.

To find reviewers:

1) There is a darwin channel at Matrix. You can ask for reviews there.

2) Seek people in nixpkgs that have merged/builds darwin PRs.

3) Find contributors using "git history" (At GitHub interface is easy) at darwin packages.

4) Remember active contributors for each ecosystem. Forward requests to those people.

How to proceed:

Add one person at a time with interval (1/day is good) until your issue is sorted.
If you were to ping too many people at once, you'd warn people unnecessarily. Since a single person is enough to handle your issue.
As we don't know who is available, we just keep adding people until someone answers.
As you are not a contributor yet you can't add 'users' for review yet. But you can message tagging. Like:

> @username Review, please?

I'll ask @wegank to review this PR since I know he is a darwin user and very active.
We should never forward the same person too many PRs, or else they get annoyed/bogged down.
Once I forwarded 200 PRs to SuperSandro2000. Poor Sandro. S2

TL-DR: Use your good sense.

File: pkgs/by-name/ch/choose-gui/package.nix
Comment (superherointj) [reply pkgs/by-name/ch/choose-gui/package.nix:7]: It is ideal to build everything from sources.
If some binary is being used, it has to be flagged as a binary package.

**Comment by superherointj** [comment]
> @superherointj I really appreciate the encouragement, it's very helpful, especially for a noob like myself! 😄

I've been there. And it's never easy, it gets better but some challenges remain, like dealing with people? But still, it is worth to keep going. The correct/ideal way of using Nix is upstreaming. If you don't, you have debt (code drifts) and lose upkeep from others. The process of upstreaming becomes easier over time. But people are still people. I should write a blog post on this. (I just need to find the time for it.)

**Comment by wegank** [comment]
Result of `nixpkgs-review pr 288931` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>choose-gui</li>
  </ul>
</details>

**Comment by wegank** [comment]
The binary works, it's just a bit sad that it's not built from source...

File: pkgs/by-name/ch/choose-gui/package.nix
Comment (heywoodlh) [reply pkgs/by-name/ch/choose-gui/package.nix]: This is resolved on the latest commit I pushed :smile:

**Comment by heywoodlh** [comment]
> The binary works, it's just a bit sad that it's not built from source...

@wegank my most recent push fixes that! :smile: Was fun dive into MacOS build tooling for me.

This PR is currently failing on this check: "This PR does not cleanly list package outputs after merging."

How do I fix that? I can't seem to find a clear explanation of this.

File: pkgs/by-name/ch/choose-gui/package.nix
Comment (superherointj) [inline pkgs/by-name/ch/choose-gui/package.nix:10]:
```suggestion
    rev = version;
```
This also works.

File: pkgs/by-name/ch/choose-gui/package.nix
Comment (superherointj) [inline pkgs/by-name/ch/choose-gui/package.nix:1]: I suggest you to use [nixfmt-rfc-style](https://search.nixos.org/packages?channel=unstable&from=0&size=50&sort=relevance&type=packages&query=nixfmt) formatter. Because is the default formatter for nixpkgs currently.

**Comment by superherointj** [comment]
>  "This PR does not cleanly list package outputs after merging."
> 
> How do I fix that? I can't seem to find a clear explanation of this.

There's an error:
> error: evaluation aborted with the following error message: 'lib.customisation.callPackageWith: Function called without required argument "system" at /home/intj/.cache/nixpkgs-review/pr-288931-2/nixpkgs/pkgs/by-name/ch/choose-gui/package.nix:1, did you mean "mystem", "systemc" or "systemd"?'


Emphasis on:

`Function called without required argument "system"`

You need to use `stdenv.hostPlatform.system` (or other options there).

File: pkgs/by-name/ch/choose-gui/package.nix
Comment (superherointj) [inline pkgs/by-name/ch/choose-gui/package.nix:1]:
```suggestion
{ lib, stdenv, fetchFromGitHub, xcbuild, coreutils, darwin }:
```

Hint: `stdenv.hostPlatform.system`

**Comment by superherointj** [comment]
> This PR is currently failing on this check: "This PR does not cleanly list package outputs after merging."

It's recommended to use [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review) to validate PRs.
You can catch problems early before submitting for review.

File: pkgs/by-name/ch/choose-gui/package.nix
Comment (heywoodlh) [reply pkgs/by-name/ch/choose-gui/package.nix:1]: Added to the latest commit! 

Also, GitHub UI/`git` workflow question: do you have suggestions on how to deal with suggested changes? I would love to commit suggestions from GitHub's UI but don't want to add an extra commit that I have to squash later -- so I've just manually added the changes and resolved the reviews. Is there a better way to work with code suggestions that I'm not aware of?

Comment (superherointj) [inline pkgs/by-name/ch/choose-gui/package.nix:1]:
```suggestion
{ lib, stdenv, fetchFromGitHub, xcbuild, coreutils, darwin }:
```
`system` is still there, you need to remove it.

Comment (superherointj) [reply pkgs/by-name/ch/choose-gui/package.nix]: I don't think there is a nice solution at GitHub UI to do that. Changes are going to be new commits. And then, I'll ask you to squash it. Ask Microsoft for a solution. :o)

Comment (heywoodlh) [reply pkgs/by-name/ch/choose-gui/package.nix]: Sorry, I forgot to push from my local machine -- latest commit fixes this :smile:

File: pkgs/by-name/ch/choose-gui/package.nix
Comment (superherointj) [reply pkgs/by-name/ch/choose-gui/package.nix:39]: Where is the changelog?

Please, don't resolve conversation while I haven't checked.

File: pkgs/by-name/ch/choose-gui/package.nix
Comment (superherointj) [inline pkgs/by-name/ch/choose-gui/package.nix:23]: I'd do something like:

```
let
  arch = with stdenv.hostPlatform;
    if isx86_64 then "x86_64"
    else if isAarch64 then "arm64"
    else throw "unsupported arch";
in
```

Notice, there are other Archs! What happens if one tries to use that? It's a flawed assumption to think everything else is `arm64`. Then what happens?

Comment (heywoodlh) [reply pkgs/by-name/ch/choose-gui/package.nix:23]: This is good feedback, although, MacOS doesn't support anything outside of x86_64+ARM64 currently. So, if an alternative arch is encountered, someone's running MacOS+Nix on a (currently) impossible hardware configuration. :smile: (Makes me curious if anyone's doing this, though, now that I think about it)

Still, though, it seems like a good idea to include, so I've added this.

One question with this: since it's impossible to run modern Nix+MacOS on anything outside of 64 bit ARM64 or x86, is there any way for me to actually test `else throw "unsupported arch"`? I.E. could I somehow override `stdenv.hostPlatform`?

Comment (superherointj) [reply pkgs/by-name/ch/choose-gui/package.nix]: No need to test that.

File: pkgs/by-name/ch/choose-gui/package.nix
Comment (heywoodlh) [reply pkgs/by-name/ch/choose-gui/package.nix:10]: Added in latest commit

File: pkgs/by-name/ch/choose-gui/package.nix
Comment (heywoodlh) [reply pkgs/by-name/ch/choose-gui/package.nix:41]: > Which last commit? It should be squashed...
Sorry, meant the most recent force-pushed/amended commit.

> Where is the changelog?
I must have switched machines and removed the changelog by force-pushing changes. Sorry, it's there now!

**Comment by superherointj** [comment]
On formatting:
* Use [nixfmt-rfc-style](https://search.nixos.org/packages?channel=unstable&from=0&size=50&sort=relevance&type=packages&query=nixfmt) formatter.
* Order items by alphabetical order whenever possible. (Attrs and Lists)

File: pkgs/by-name/ch/choose-gui/package.nix
Comment (heywoodlh) [reply pkgs/by-name/ch/choose-gui/package.nix]: Added in most recent commit

File: pkgs/by-name/ch/choose-gui/package.nix
Comment (superherointj) [inline pkgs/by-name/ch/choose-gui/package.nix:17]:
```suggestion
    coreutils
    darwin.apple_sdk.frameworks.Cocoa
    xcbuild
```
(Whenever possible) follow alphabetical order.

Comment (heywoodlh) [reply pkgs/by-name/ch/choose-gui/package.nix]: Done, thanks!

**Comment by heywoodlh** [comment]
> On formatting:
Use [nixfmt-rfc-style](https://search.nixos.org/packages?channel=unstable&from=0&size=50&sort=relevance&type=packages&query=nixfmt) formatter.

Just updated my vim flake to use nixfmt-rfc-style with ALE instead of nixfmt classic :smile:

Is there something specific in the styling I should add? I'm not seeing any complaints from nixfmt-rfc-style in my vim setup.

**Comment by superherointj** [comment]
RFC has chosen `nixfmt-rfc-style`.
Avoid `alejandra` or `nixfmt(-classic)`.

`nixfmt-rfc-style` is available from `24.05` or `unstable`. `23.11` is missing it.

Use it manually and check. I think the difference will be clear. Particularly on parameters.

File: pkgs/by-name/ch/choose-gui/package.nix
Comment (superherointj) [inline pkgs/by-name/ch/choose-gui/package.nix:37]: Can you move this `let arch = ...; in` to line 10 (before `stdenv.mkDerivation rec {`)?

I understand you wanted to keep it scoped right. But in nixpkgs this is unusual, these kind of 'let .. in' are usually kept before `stdenv.mkDerivation`.
To let the formatter format it right, I suggest you make it a single line and the formatter adjust it for you. The formatter is not optionated on break lines.

**Comment by superherointj** [comment]
> Let me know if there's anything else 😄

I'm mostly happy with package now.

I have provided you with the most advice/input I could to advance things.

Now you just need a darwin maintainer to have it merged. I suggest to re-ping our friend wegank to check if he approves/merges this.

File: pkgs/by-name/ch/choose-gui/package.nix
Comment (wegank) [inline pkgs/by-name/ch/choose-gui/package.nix:24]: I think `xcbuild` goes to `nativeBuildInputs`. Try setting `strictDeps = true;` and see if this is the case.

File: pkgs/by-name/ch/choose-gui/package.nix
Comment (wegank) [inline pkgs/by-name/ch/choose-gui/package.nix:40]:
```suggestion
      xcodebuild -arch ${stdenv.hostPlatform.darwinArch} -configuration Release SYMROOT="./output" build
```

And drop `let arch = ... in`.

File: pkgs/by-name/ch/choose-gui/package.nix
Comment (wegank) [inline pkgs/by-name/ch/choose-gui/package.nix:22]:
```suggestion
```

`coreutils` should be part of stdenv.

**Comment by superherointj** [comment]
ping @heywoodlh

File: pkgs/by-name/ch/choose-gui/package.nix
Comment (heywoodlh) [reply pkgs/by-name/ch/choose-gui/package.nix]: Resolved on most latest push!

File: pkgs/by-name/ch/choose-gui/package.nix
Comment (heywoodlh) [reply pkgs/by-name/ch/choose-gui/package.nix]: Resolved on most latest push!

File: pkgs/by-name/ch/choose-gui/package.nix
Comment (heywoodlh) [reply pkgs/by-name/ch/choose-gui/package.nix]: You were correct, when I had `strictDeps = true;` and left `xcbuild` in `buildInputs` it failed -- but it worked when I moved `xcbuild` to `nativeBuildInputs`.

Resolved on latest commit!

Comment (heywoodlh) [reply pkgs/by-name/ch/choose-gui/package.nix]: (I haven't been able to find super clear definitions differentiating between `buildInputs` and `nativeBuildInputs` -- would love any pointers to documentation that clearly describe them if possible so I can better understand)

Comment (superherointj) [reply pkgs/by-name/ch/choose-gui/package.nix]: > (I haven't been able to find super clear definitions differentiating between `buildInputs` and `nativeBuildInputs` -- would love any pointers to documentation that clearly describe them if possible so I can better understand)

- `buildInputs` are for runtime dependencies, it's going to be included in the final package, because the runtime requires it.
- `nativeBuildInputs` are for build time dependencies, usually tools that aren't needs in the final package but are used during build.

This distinction is important when doing cross-compilation. The dependencies (usually tools) used for building are going to run in the `host architecture`, while dependencies included in the output package (which runtime/buildInputs is part of) will be in the `target architecture`.

So the question you should be asking is:
* Is this dependency needed for the application to run (runtime)?
  - If yes, `buildInputs`.
  - If no, `nativeBuildInputs`.

If it is build tools, like `cmake`, it's `nativeBuildInputs`.

**Comment by superherointj** [comment]
Result of `nixpkgs-review pr 288931` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>choose-gui</li>
  </ul>
</details>

**Review by superherointj — APPROVED** [review]

**Comment by superherointj** [comment]
@heywoodlh Your contributions will get easier over time as you learn the patterns. You'll see.
Nixpkgs is a harsh place, but eventually you'll learn how to endure it better. :-)
Welcome to ~~the jungle~~ nixpkgs!

**Pull request merged**
